cmd/ctr/checkpoint.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"bytes"
+	gocontext "context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"runtime"
+
+	"github.com/containerd/containerd/api/services/execution"
+	"github.com/containerd/containerd/api/types/descriptor"
+	"github.com/containerd/containerd/archive"
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/urfave/cli"
+)
+
+var checkpointCommand = cli.Command{
+	Name:  "checkpoint",
+	Usage: "checkpoint a container",
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "id",
+			Usage: "id of the container",
+		},
+		cli.BoolFlag{
+			Name:  "exit",
+			Usage: "stop the container after the checkpoint",
+		},
+		cli.BoolFlag{
+			Name:  "binds",
+			Usage: "checkpoint bind mounts with the checkpoint",
+		},
+	},
+	Action: func(context *cli.Context) error {
+		var (
+			id  = context.String("id")
+			ctx = gocontext.Background()
+		)
+		if id == "" {
+			return errors.New("container id must be provided")
+		}
+		containers, err := getExecutionService(context)
+		if err != nil {
+			return err
+		}
+		content, err := getContentStore(context)
+		if err != nil {
+			return err
+		}
+
+		var spec specs.Spec
+		if context.Bool("binds") {
+			info, err := containers.Info(ctx, &execution.InfoRequest{
+				ID: id,
+			})
+			if err != nil {
+				return err
+			}
+			if err := json.Unmarshal(info.Spec.Value, &spec); err != nil {
+				return err
+			}
+		}
+		checkpoint, err := containers.Checkpoint(ctx, &execution.CheckpointRequest{
+			ID:   id,
+			Exit: context.Bool("exit"),
+		})
+		if err != nil {
+			return err
+		}
+		var additionalDescriptors []*descriptor.Descriptor
+		if context.Bool("binds") {
+			if additionalDescriptors, err = checkpointBinds(ctx, &spec, content); err != nil {
```

Comment:
if `context.Bool("exit")` is false, shouldn't this be done after pausing the container?